### ISSUE TITLE
Header Bar stuck in progress state

### ISSUE THREAD

**Issue by adrianq** [dhis2/d2-ui#94]
We are building a dhis2 web app and we are using the [app-skeleton](https://github.com/dhis2/app-skeleton/) as a starting point. Everything seems to work fine when we used the [npm d2-ui](https://www.npmjs.com/package/d2-ui).

However, as we need to customise some components we are linking d2-ui locally with our app and building it. Everything seems to be OK. Our customised components are loaded and everything works as expected however the header bar stays like this (progress bar):
![screenshot from 2016-11-25 13 19 22](https://cloud.githubusercontent.com/assets/6850223/20626460/d10b28ee-b311-11e6-8d5c-069489f97ab2.png)

We have tried to compile d2-ui using npm run build and webpack -p without succeed. We are using v24 but I give it a try to v25 a couple of weeks ago and couldn't make it work. I was looking for documentation but couldn't find anything. any ideas? am I missing something?

cc @josemp10

**Comment by nicolayr** [comment]
This can happen if your app depends on a different version of d2 than the version of d2-ui you're using does.

That's a mouthful, so allow me to illustrate with some beautiful ascii art:

Good scenario:
       
app --> d2 <-- d2-ui

Bad scenario:

d2 <-- app --> d2-ui --> _different_ d2

And a more detailed explanation:

The header bar is designed to wait for the main app to initialize d2, and then use that instance to fetch what it needs from the API. However, if d2-ui and your app depend on different versions of d2, they will each get their own, separate instance of d2. Therefore the d2 instance used by d2-ui will never be initialized, and the loading bar never goes away.

In short, this should be solved if you update both d2 and d2-ui to the latest version.

**Comment by adrianq** [comment]
Thanks for your quick response @nicolayr ! The problem was the d2 library, as you mentioned, however I wonder how you usually develop.

In my case, I was using `npm link` in order to link my web app to the modified d2-ui. This was not working (header stays in progress forever) as d2-ui could find d2 inside its own node_modules folder. If I delete d2-ui/node_modules and I copy d2-ui folder inside my web_app/node_modules everything works as expected. d2-ui can't find dependencies inside its own node_modules so it goes up to its parent folder (my webapp) to look for these dependencies. Therefore, web_app --> d2 <-- d2-ui is true.

This is not a big deal as I can develop using npm link with a forever-in-progress header and just by the time I move everything to production do the trick-hack.... but is this the right way? Anyway, thank you very much for your help!

**Comment by nicolayr** [comment]
I did use `npm link` in the past, but I eventually gave up because it gave me too many issues. What works well for me is to build d2-ui and then install my local copy in the app. I'm using this alias to make it easier:

`alias d2UiBuild="pushd <d2-ui-location>; npm run build-only; popd; npm install <d2-ui-location>"`

And while working on stuff in the app I do: `d2UiBuild && npm start`
